### DISCUSSION TITLE
Running flatcar on WSL-2

### DISCUSSION THREAD

**Discussion by mwoodpatrick** [flatcar/Flatcar#1257] [Q&A]
I would like to try and install & test flatcar on WSL-2. Is this possible any pointers on how to do this or to related docs would be much appreciated.

**Comment by donno** [comment]
I wasn't able to get it working, but t may provide a starting point for others (i.e here are the pointers I have)

To import a Linux distribution into WSL-2, you can use a tar containing the root filesystem..
The main requirement for the tarball to be considered a valid distribution is is presence of the `/etc` and `/bin/sh` (which can be linked to `/usr/bin/bash`. WSL can also read `/etc/os-release` but that seems to be optional.

The `flatcar-container.tar.gz` does not honour the above requirement, and adding them wasn't enough to get a working system.

I'm unfamiliar with flatcar as I was liking to try and test it out as well.
The next approach I tried was converting `flatcar_production_image_sysext.squashfs` to the rootfs.

In Ubuntu Distro or container (I used a container)
* apt-get install -y squashfs-tools-ng
* sqfs2tar /host/flatcar_production_image_sysext.squashfs > flatcar_production_image_sysext.tar
* cp flatcar_production_image_sysext.tar /host
There were issues trying to redirect to a file directly on host as it would get stuck, so easier to write to the image.

Now that tar can be imported.
`wsl --import FlatCar %PROGRAMDATA%\WSLDistroStorage\flatcar flatcar_production_image_sysext.tar`

```
$ wsl -d FlatCar --cd /
Update Strategy: No Reboots
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
MyPC / #
MyPC / # cat /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=4152.2.2
VERSION_ID=4152.2.2
BUILD_ID=2025-03-17-1639
~~~

MyPC / # vi /etc/wsl.conf
# Add the following within that file.
[boot]
systemd=true
```
Restart wsl (wsl --shutdown)

Now systemd will start within that distro.

```
Update Strategy: No Reboots
Failed Units: 2
  audit-rules.service
  systemd-sysext.service
```
However it still failed:
```
$ journalctl --catalog --pager-end --unit systemd-sysext.service

Jun 01 15:42:07 MyPC systemd-sysext[259]: Failed to read metadata for image containerd-flatcar: Invalid argument
Jun 01 15:42:07 MyPC systemd[1]: systemd-sysext.service: Main process exited, code=exited, status=1/FAILURE
```

Hopefully, this helps someone else. That said, I'm not convinced the sysext squashfs is the correct starting point.

The other thing I tried was importing the vhd/vhdx hard drive images for the Hyper-V distribution using `--import-in-place` which takes a vhd and makes it the base-image of the distro however it claimed both were corrupt, so it seems it doesn't like how they are partitioned with an EFI boot partition. I didn't try to look into WSL's code to figure out what the cause of that error was in this case.
